### PR TITLE
[codex] Fix Anthropic tool result handling

### PR DESCRIPTION
**Pull request by sean1588**
## Summary

Fixes the Anthropic model client request shaping for tool-use conversations and tightens Anthropic-specific request validation behavior.

## What changed

- Groups consecutive internal tool-result messages into a single Anthropic `user` message containing multiple `tool_result` blocks.
- Adds placeholder `tool_result` blocks when a previous assistant tool call is present in history without a recorded result, matching the OpenAI-compatible converter’s recovery behavior.
- Clamps Anthropic extended-thinking budgets to the provider minimum of 1024 tokens and omits `thinking` when the available output budget is too small.
- Downgrades `toolChoice: "required"` to Anthropic `auto` when extended thinking is enabled, avoiding an incompatible request shape.
- Adds focused mock-client coverage for parallel tool calls, interrupted histories, tool choice, structured output, and thinking budget behavior.

## Why

Anthropic’s Messages API is stricter than OpenAI-compatible chat formats: `tool_use` blocks need corresponding `tool_result` blocks immediately afterward in the next user message. The previous converter emitted each tool result as its own user message, which is risky for parallel tool calls and can lead to API 400s once the client is exercised against real Anthropic keys.

## Validation

- `PATH=/root/.nvm/versions/node/v22.16.0/bin:$PATH node --test --import tsx tests/model-client-anthropic-structured-output.test.ts`
- `PATH=/root/.nvm/versions/node/v22.16.0/bin:$PATH node --test --import tsx tests/model-client-*.test.ts`
- `PATH=/root/.nvm/versions/node/v22.16.0/bin:$PATH npm run build`
- `PATH=/root/.nvm/versions/node/v22.16.0/bin:$PATH npm run lint --workspace=packages/agent-sdk`

Note: I could not run a live Anthropic API smoke test because no `ANTHROPIC_API_KEY` is configured yet.

### CODE REVIEW

File: packages/agent-sdk/src/model/client.ts [hunk 1]
```diff
@@ -753,10 +786,27 @@ export class AnthropicModelClient implements ModelClient {
...
+        params.reasoningMaxTokens > 0 &&
+        params.reasoningMaxTokens < ANTHROPIC_MIN_THINKING_BUDGET_TOKENS
+          ? 0
+          : ANTHROPIC_MIN_THINKING_BUDGET_TOKENS;
+    }
+    if (cappedBudget >= params.maxTokens) {
+      cappedBudget = params.maxTokens - 1;
+    }
+    if (cappedBudget < ANTHROPIC_MIN_THINKING_BUDGET_TOKENS) {
+      cappedBudget = 0;
```
Comment (sean1588) [inline packages/agent-sdk/src/model/client.ts]: **Major (form, not a bug):** This clamp mutates `cappedBudget` in three steps — set floor (with a nested ternary), apply ceiling (`maxTokens - 1`), then re-check the floor *again* because the ceiling step can re-violate it. The valid-range invariant `[1024, maxTokens)` can't be read off the code; you have to simulate three mutations, and the second floor-check exists only because of mutation ordering (a self-inflicted edge case).

It is functionally correct (traced: `maxTokens=1024 → 1023 → 0`, no leak). Suggestion: extract a pure `clampThinkingBudget(requested, maxTokens)` that states the invariant once — if `requested < MIN` or `maxTokens - 1 < MIN` → `0`; else `Math.min(requested, maxTokens - 1)`. Kills the double floor-check.

File: packages/agent-sdk/src/model/client.ts [hunk 2]
```diff
@@ -82,21 +110,27 @@ function toAnthropicMessages(
...
-        {
-          type: "tool_result",
-          tool_use_id: message.toolCallId,
-          content: message.content,
-        },
-      ],
+    const pendingToolCallIndex = pendingToolCalls.findIndex(
+      (toolCall) => toolCall.id === message.toolCallId,
+    );
+    if (pendingToolCallIndex === -1) {
```
Comment (sean1588) [inline packages/agent-sdk/src/model/client.ts:125]: **Major (test gap):** This orphan-drop branch — a `tool` message whose id matches no pending call — is the silent-skip path, and nothing tests it. A regression that pushed the orphan anyway would emit a `tool_result` with an undeclared `tool_use_id`: exactly the malformed shape that causes Anthropic 400s, i.e. the bug class this PR exists to prevent.

Suggestion: add a test with an assistant declaring one call (`tu_1`) plus a `tool` message for a non-existent `tu_99`; assert the orphan is dropped and only `tu_1` survives.

File: packages/agent-sdk/src/model/client.ts [hunk 3]
```diff
@@ -50,9 +53,33 @@ function toAnthropicMessages(
   messages: SessionMessage[],
 ): Anthropic.MessageParam[] {
   const converted: Anthropic.MessageParam[] = [];
+  let pendingToolCalls: ToolCall[] = [];
+  let pendingToolResults: Anthropic.ContentBlockParam[] = [];
+
+  const flushPendingToolResults = () => {
```
Comment (sean1588) [inline packages/agent-sdk/src/model/client.ts]: **Minor (abstraction):** After this PR `toAnthropicMessages` and `toOpenAICompatibleMessages` (`client.ts:424`) share the identical recovery policy — buffer pending calls, skip orphans via `findIndex === -1 → continue`, synthesize `MISSING_TOOL_RESULT_CONTENT` placeholders on flush at boundaries. Only the wire encoding differs (Anthropic batches into one `user` message; OpenAI emits one `tool` message per result). They are now the same *concept*, not just the same shape — if the orphan/placeholder policy changes, one converter silently diverges.

Suggestion: extract a shared traversal that yields a normalized event sequence and let each provider render its wire format. If deferred, at minimum comment both functions that they must stay in lockstep.

File: packages/agent-sdk/src/model/client.ts [hunk 4]
```diff
@@ -82,21 +110,27 @@ function toAnthropicMessages(
...
-        },
-      ],
+    const pendingToolCallIndex = pendingToolCalls.findIndex(
+      (toolCall) => toolCall.id === message.toolCallId,
+    );
+    if (pendingToolCallIndex === -1) {
+      continue;
+    }
+
+    pendingToolResults.push({
```
Comment (sean1588) [inline packages/agent-sdk/src/model/client.ts:129]: **Minor (test gap):** Out-of-order tool results aren't pinned. The code drains `pendingToolCalls` by id (`findIndex`+`splice`), so reverse-order results work — but the only grouping test delivers `tu_1` then `tu_2` in order. Someone "optimizing" to `pendingToolCalls.shift()` would pass all current tests yet emit spurious placeholders.

Suggestion: add a grouping-test variant delivering `tu_2` before `tu_1`; assert both real results are emitted with no placeholder content.

File: packages/agent-sdk/src/model/client.ts [hunk 5]
```diff
@@ -753,10 +786,27 @@ export class AnthropicModelClient implements ModelClient {
...
-    const cappedBudget =
+    let cappedBudget =
       params.reasoningMaxTokens !== undefined && params.reasoningMaxTokens > 0
         ? Math.min(effortBudget || params.reasoningMaxTokens, params.reasoningMaxTokens)
         : effortBudget;
+    if (
+      cappedBudget > 0 &&
+      cappedBudget < ANTHROPIC_MIN_THINKING_BUDGET_TOKENS
+    ) {
+      cappedBudget =
```
Comment (sean1588) [inline packages/agent-sdk/src/model/client.ts]: **Minor (test gap):** The deliberate asymmetry encoded here — an *explicit* `reasoningMaxTokens < 1024` means "don't think" (→ `0`), vs. an effort-derived small budget means "think at the minimum" (→ `MIN`) — has no coverage. The existing test hits the `→ MIN` arm via `reasoningEffort: "minimal"`; the `→ 0` arm is never exercised.

Suggestion: add a test with `reasoningMaxTokens: 512`, `maxTokens: 8192`; assert `thinking === undefined`.

File: packages/agent-sdk/src/model/client.ts [hunk 6]
```diff
@@ -643,9 +674,11 @@ function effortToBudgetFraction(effort: ReasoningEffort): number {
 function mapAnthropicToolChoice(
   choice: ToolChoice | undefined,
   toolCount: number,
+  thinkingEnabled = false,
 ): { type: "auto" } | { type: "any" } | { type: "none" } | null {
   if (choice === undefined) return null;
   if (choice === "required" && toolCount === 0) return { type: "auto" };
+  if (choice === "required" && thinkingEnabled) return { type: "auto" };
```
Comment (sean1588) [inline packages/agent-sdk/src/model/client.ts]: **Minor (readability):** This branch reads as an arbitrary special case but is actually a provider constraint (Anthropic rejects forced `tool_choice` with extended thinking). The conditional is fine where it lives, but its reason is invisible.

Suggestion: add a one-line comment stating the constraint. Optionally collapse with the zero-tools case: `if (choice === "required" && (toolCount === 0 || thinkingEnabled)) return { type: "auto" };`

File: packages/agent-sdk/src/model/client.ts [hunk 7]
```diff
@@ -82,21 +110,27 @@ function toAnthropicMessages(
...
+
+    pendingToolResults.push({
+      type: "tool_result",
+      tool_use_id: message.toolCallId,
+      content: message.content,
     });
+    pendingToolCalls.splice(pendingToolCallIndex, 1);
   }
 
+  flushPendingToolResults();
```
Comment (sean1588) [inline packages/agent-sdk/src/model/client.ts]: **Nit (consistency):** This flush helper is named `flushPendingToolResults`, but the analogous function in `toOpenAICompatibleMessages` is `flushMissingToolResults`. Different names for the mirrored function obscure the parallel structure — and since this version also synthesizes placeholders for missing calls, "missing" arguably fits better. Suggestion: align the two names.

File: packages/agent-sdk/src/model/client.ts [hunk 8]
```diff
@@ -770,6 +820,7 @@ export class AnthropicModelClient implements ModelClient {
     const anthropicToolChoice = mapAnthropicToolChoice(
       params.toolChoice,
       toolsForRequest.length,
+      cappedBudget > 0,
```
Comment (sean1588) [inline packages/agent-sdk/src/model/client.ts]: **Nit (consistency):** Neighboring code uses named intermediates (`effortBudget`, `cappedBudget`); the thinking flag is passed here as a bare expression. Suggestion: `const thinkingEnabled = cappedBudget > 0;` then pass `thinkingEnabled` — makes the budget↔toolChoice relationship explicit at the call site.

File: packages/agent-sdk/src/model/client.ts [hunk 9]
```diff
@@ -50,9 +53,33 @@ function toAnthropicMessages(
   messages: SessionMessage[],
 ): Anthropic.MessageParam[] {
   const converted: Anthropic.MessageParam[] = [];
+  let pendingToolCalls: ToolCall[] = [];
+  let pendingToolResults: Anthropic.ContentBlockParam[] = [];
```
Comment (sean1588) [inline packages/agent-sdk/src/model/client.ts:59]: **Nit (docs):** This extra `pendingToolResults` buffer (vs. the OpenAI converter, which pushes results directly) exists only because Anthropic requires consecutive `tool_result` blocks grouped in a single `user` message — the one legitimate divergence between the converters. Suggestion: a one-line comment noting why results are buffered rather than pushed immediately.

**Review by sean1588 — COMMENTED** [review]
Automated structural/implementation/consistency/test review. The fix is functionally correct (edge cases traced, all 11 tests pass) — comments below are about form, a duplicated concept across the two converters, and two real coverage gaps. None block merge.
